clang-tools-extra/clang-tidy/cppcoreguidelines/AvoidCapturingLambdaCoroutinesCheck.cpp

```diff
@@ -0,0 +1,43 @@
+//===--- AvoidCapturingLambdaCoroutinesCheck.cpp - clang-tidy -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "AvoidCapturingLambdaCoroutinesCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang {
+namespace tidy {
+namespace cppcoreguidelines {
+
+void AvoidCapturingLambdaCoroutinesCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(lambdaExpr().bind("lambda"), this);
+}
+
+void AvoidCapturingLambdaCoroutinesCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *Lambda = Result.Nodes.getNodeAs<LambdaExpr>("lambda");
+  if (!Lambda) {
+    return;
+  }
+
+  const auto *Body = dyn_cast<CoroutineBodyStmt>(Lambda->getBody());
+  if (!Body) {
+    return;
+  }
+
+  if (Lambda->captures().empty()) {
+    return;
+  }
+
+  diag(Lambda->getBeginLoc(), "found capturing coroutine lambda");
```

Comment:
Is it worth checking that the captures are used after a suspension point, perhaps dependant on whether `suspend_never` or suspend_always is in use? Or are we happy with he false positives?

Reply:
this is cool. so is void so that this is the checker that does *something* and `diag` does all the real work? (printing, registering failed checker, etc)

Reply:
i'm happy with it--it's kind of like a unused-variable-warning in that case. also, if you do implicit captures like & and = then this check already doesn't fire if you don't actually capture anything.

Reply:
> this is cool. so is void so that this is the checker that does something and diag does all the real work? (printing, registering failed checker, etc)

mm, i don't think `diag` is doing any checking. it's a visitor pattern over the ast.